SPEC.md

```diff
@@ -3352,6 +3469,162 @@ struct Invalid {
 }
 ```
 
+## Enum Definition
+
+An `enum` is an enumerated type. Enums enable the creation of types that represent closed sets of alternatives (called "variants") that are semantically valid in a specific context. Once defined, an `enum` type can be used as the type of a declaration like any other type. However, new variants of an `enum` cannot be created. Instead, a declaration having an `enum` type must be assigned one of the variants created as part of the `enum`'s definition.
+
+An enum definition is a top-level WDL element, meaning it is defined at the same level as tasks, workflows, and structs, and it cannot be defined within a task or workflow body. An enum is defined using the `enum` keyword, followed by a name that is unique within the WDL document, and a body containing a comma-delimited list of varaints in braces (`{}`).
+
+```wdl
+enum Color {
+  Red,
+  Blue,
+  Green
+}
+```
+
+An enum can be thought of as a closed type with a fixed set of instances. The `enum` keyword creates both a type (that can be used in declarations) and a global namespace containing the enum's variants. For example, `Color.Red` refers to a specific instance of the `Color` enum type.
+
+Unlike structs, it is not possible to create new instances of an `enum` outside of the enum's definition. An enum value can only be one of the variants defined in the enum's declaration.
+
+### Enum Usage
+
+An `enum`'s variants are [accessed](#member-access) using a `.` to separate the variant name from the `enum`'s identifier.
+
+A declaration with an `enum` type can only be initialized by referencing a variant directly or by assigning it to the value of another declaration of the same `enum` type.
+
+Two enum values can be tested for equality (i.e., using `==` or `!=`). To be equal, two enum values must be the same variant of the same `enum` type. Enum variants are not ordered, so they cannot be compared (i.e., using `>`, `>=`, `<`, `<=`).
+
+When an enum value is serialized (for example, in a command section using string interpolation), it is serialized to its inner value. To access the stringified variant name, use the [`name()`](#name) standard library function.
+
+An `enum` cannot be coerced to or from any other type. However, an enum value can be [serialized to/deserialized from a `String`]() or to/from [JSON]().
+
+```wdl
+version 1.2
+enum Pet {
+  Cat,
+  Mouse,
+  Bird
+}
+enum ComputerDevice {
+  Mouse,
+  Keyboard,
+  Monitor
+}
+task compare_enum_types {
+  input {
+    Pet? pet
+  }
+
+  Pet my_pet = select_first([pet, Pet.Mouse])
+  Array[String] all_pet_names = names(Pet)
+
+  command <<<
+    echo "There are ~{length(all_pet_names)} kinds of pet: ~{sep(", ", all_pet_names)}"
+    echo "I have a pet ~{name(my_pet)}"
+    echo "My pet is a ~{my_pet}"
+  >>>
+
+  output {
+    Boolean different_types = Pet.Mouse != ComputerDevice.Mouse
```

Comment:
@claymcleod what are the rules around equality? Would this always evaluate to `true` or Is this effectively equivalent to `value(Pet.Mouse) != value(ComputerDevice.Mouse)` which would be false?

Reply:
In my mind, it is not equivalent to `value(Pet.Mouse) != value(ComputerDevice.Mouse)`, because the values within each variant do not constitute the entire identity of the type (though you can get the inner values out of variants and compare equality that way, if that's what you desire). Comparing enum variants themselves is much more akin to comparing two totally different types.

We treat structs the same way: consider two structs, each with a single member `foo: String` with a value `"bar"`; though the constituent data within the structs are identical, the fact that they are typed different necessitates the inequality because the type is part of the identity.